### PR TITLE
[Feature] Add RiRoIAlignRotated CUDA op for rotated detection.

### PR DESCRIPTION
**Pull request by zytx121**
Add RiRoIAlignRotated CUDA ops for rotated detection.

```
@InProceedings{han2021ReDet,
    author    = {Han, Jiaming and Ding, Jian and Xue, Nan and Xia, Gui-Song},
    title     = {ReDet: A Rotation-equivariant Detector for Aerial Object Detection},
    booktitle = {Proceedings of the IEEE/CVF Conference on Computer Vision and Pattern Recognition (CVPR)},
    month     = {June},
    year      = {2021},
    pages     = {2786-2795}
}

```

## Modification

Several files in mmcv/ops folder.

## BC-breaking (Optional)

No.

## Use cases (Optional)

from mmcv.ops import RiRoIAlignRotated

### CODE REVIEW

**Comment by zhouzaida** [comment]
Please @AllentDan have a look.

File: tests/test_ops/test_riroi_align_rotated.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        np_feature, dtype=torch.float, device='cuda', requires_grad=True)
+    rois = torch.tensor(np_rois, dtype=torch.float, device='cuda')
+    froipool = RiRoIAlignRotated((pool_h, pool_w), spatial_scale, sample_num,
+                                 nOrientation, clockwise)
+    gradcheck(froipool, (x, rois), eps=1e-3, atol=1e-3)
+
+
+@pytest.mark.skipif(
+    not torch.cuda.is_available(), reason='requires CUDA support')
+def test_roialign_rotated_allclose():
```
Comment (grimoire) [inline tests/test_ops/test_riroi_align_rotated.py:62]: Is this the real gradcheck?
And this test failed on my device.

Comment (zytx121) [reply tests/test_ops/test_riroi_align_rotated.py:62]: Sorry, I mistakenly believe that there is a CUDA environment in CI. I'll test it locally.

Comment (zytx121) [reply tests/test_ops/test_riroi_align_rotated.py:62]: Please test again now. Thx.

**Review by grimoire — APPROVED** [review]
LGTM

File: mmcv/ops/riroi_align_rotated.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+class RiRoIAlignRotated(nn.Module):
+    """Rotation-invariant RoI align pooling layer for rotated proposals.
+
+    It accepts a feature map of shape (N, C, H, W) and rois with shape
+    (n, 6) with each roi decoded as (batch_index, center_x, center_y,
+    w, h, angle). The angle is in radian.
+
+    The details are described in the paper `ReDet: A Rotation-equivariant
+    Detector for Aerial Object Detection  <https://arxiv.org/abs/2103.07733>`.
```
Comment (zhouzaida) [inline mmcv/ops/riroi_align_rotated.py:90]:
```suggestion
    Detector for Aerial Object Detection  <https://arxiv.org/abs/2103.07733>`_.
```

File: mmcv/ops/riroi_align_rotated.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+    Args:
+        out_size (tuple): h, w
+        spatial_scale (float): scale the input boxes by this number
+        sample_num (int): number of inputs samples to take for each
+            output sample. 0 to take samples densely for current models.
+        nOrientation (int): number of oriented channels.
```
Comment (zhouzaida) [inline mmcv/ops/riroi_align_rotated.py:97]: suggest renaming the `nOrientation` to `num_orientations`

File: mmcv/ops/riroi_align_rotated.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    (n, 6) with each roi decoded as (batch_index, center_x, center_y,
+    w, h, angle). The angle is in radian.
+
+    The details are described in the paper `ReDet: A Rotation-equivariant
+    Detector for Aerial Object Detection  <https://arxiv.org/abs/2103.07733>`.
+
+    Args:
+        out_size (tuple): h, w
+        spatial_scale (float): scale the input boxes by this number
+        sample_num (int): number of inputs samples to take for each
```
Comment (zhouzaida) [inline mmcv/ops/riroi_align_rotated.py:95]: suggest renaming the `sample_num` to `num_samples`

File: mmcv/ops/riroi_align_rotated.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+    It accepts a feature map of shape (N, C, H, W) and rois with shape
+    (n, 6) with each roi decoded as (batch_index, center_x, center_y,
+    w, h, angle). The angle is in radian.
+
+    The details are described in the paper `ReDet: A Rotation-equivariant
+    Detector for Aerial Object Detection  <https://arxiv.org/abs/2103.07733>`.
+
+    Args:
+        out_size (tuple): h, w
```
Comment (zhouzaida) [inline mmcv/ops/riroi_align_rotated.py:93]: please describe the meaning of `out_size`.

File: mmcv/ops/riroi_align_rotated.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+                sample_num=0,
+                nOrientation=8,
+                clockwise=False):
+        if isinstance(out_size, int):
+            out_h = out_size
+            out_w = out_size
+        elif isinstance(out_size, tuple):
+            assert len(out_size) == 2
+            assert isinstance(out_size[0], int)
+            assert isinstance(out_size[1], int)
```
Comment (zhouzaida) [inline mmcv/ops/riroi_align_rotated.py:28]:
```suggestion
        elif is_tuple_of(out_size, int):
            assert len(out_size) == 2
```

https://github.com/open-mmlab/mmcv/blob/e4b5348ebf407f0886bc11caa7ca74d2fbf693ac/mmcv/utils/misc.py#L160

File: mmcv/ops/riroi_align_rotated.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+            out_h, out_w = out_size
+        else:
+            raise TypeError(
+                '"out_size" must be an integer or tuple of integers')
```
Comment (zhouzaida) [inline mmcv/ops/riroi_align_rotated.py:32]:
```suggestion
                f'"out_size" should be an integer or tuple of integers, but got {out_size}')
```

File: mmcv/ops/riroi_align_rotated.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        ctx.spatial_scale = spatial_scale
+        ctx.sample_num = sample_num
+        ctx.nOrientation = nOrientation
+        ctx.clockwise = clockwise
+        ctx.save_for_backward(rois)
+        ctx.feature_size = features.size()
+
+        batch_size, num_channels, data_height, data_width = features.size()
```
Comment (zhouzaida) [inline mmcv/ops/riroi_align_rotated.py:40]:
```suggestion
        batch_size, num_channels, feature_height, feature_width = features.size()
```

File: mmcv/ops/riroi_align_rotated.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    @staticmethod
+    def backward(ctx, grad_output):
+        feature_size = ctx.feature_size
+        spatial_scale = ctx.spatial_scale
+        nOrientation = ctx.nOrientation
+        clockwise = ctx.clockwise
+        sample_num = ctx.sample_num
+        rois = ctx.saved_tensors[0]
+        assert feature_size is not None
+        batch_size, num_channels, data_height, data_width = feature_size
```
Comment (zhouzaida) [inline mmcv/ops/riroi_align_rotated.py:60]:
```suggestion
        batch_size, num_channels, feature_height, feature_width = feature_size
```

**Comment by zhouzaida** [comment]
We also need to update those ops in https://github.com/open-mmlab/mmcv/blob/master/docs/en/understand_mmcv/ops.md and https://github.com/open-mmlab/mmcv/blob/master/docs/zh_cn/understand_mmcv/ops.md

**Review by AllentDan — APPROVED** [review]
LGTM

File: mmcv/ops/csrc/common/cuda/riroi_align_rotated_cuda_kernel.cuh [hunk 10]
```diff
@@ -0,0 +1,246 @@
+// Modified from
+// https://github.com/csuhan/ReDet/blob/master/mmdet/ops/riroi_align/src/riroi_align_kernel.cu
+#ifndef RIROI_ALIGN_ROTATED_CUDA_KERNEL_CUH
+#define RIROI_ALIGN_ROTATED_CUDA_KERNEL_CUH
+
+#include <float.h>
+#ifdef MMCV_WITH_TRT
+#include "common_cuda_helper.hpp"
```
Comment (zhouzaida) [inline mmcv/ops/csrc/common/cuda/riroi_align_rotated_cuda_kernel.cuh:8]: @grimoire , is it necessary to include the `common_cuda_helper.hpp`  here?

Comment (grimoire) [reply mmcv/ops/csrc/common/cuda/riroi_align_rotated_cuda_kernel.cuh:8]: Nope, we do not plan to add TensorRT support for now.

**Review by zhouzaida — APPROVED** [review]

File: mmcv/ops/csrc/pytorch/cuda/riroi_align_rotated_cuda.cu [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    const int num_samples, const bool clockwise, const int channels,
+    const int height, const int width, const int num_rois,
+    const int pooled_height, const int pooled_width, const int num_orientations,
+    at::Tensor output) {
+  const int output_size =
+      num_rois * pooled_height * pooled_width * channels * num_orientations;
+  at::cuda::CUDAGuard device_guard(features.device());
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+      features.scalar_type(), "RiROIAlignRotatedLaucherForward", ([&] {
```
Comment (zhouzaida) [inline mmcv/ops/csrc/pytorch/cuda/riroi_align_rotated_cuda.cu:16]:
```suggestion
      features.scalar_type(), "riroi_align_rotated_forward_cuda_kernel", ([&] {
```

File: mmcv/ops/csrc/pytorch/cuda/riroi_align_rotated_cuda.cu [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    const int num_samples, const bool clockwise, const int channels,
+    const int height, const int width, const int num_rois,
+    const int pooled_height, const int pooled_width, const int num_orientations,
+    at::Tensor bottom_grad) {
+  const int output_size =
+      num_rois * pooled_height * pooled_width * channels * num_orientations;
+  at::cuda::CUDAGuard device_guard(top_grad.device());
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+      top_grad.scalar_type(), "RiROIAlignRotatedLaucherBackward", ([&] {
```
Comment (zhouzaida) [inline mmcv/ops/csrc/pytorch/cuda/riroi_align_rotated_cuda.cu:42]:
```suggestion
      top_grad.scalar_type(), "riroi_align_rotated_backward_cuda_kernel", ([&] {
```

**Review by zhouzaida — APPROVED** [review]

File: mmcv/ops/riroi_align_rotated.py [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+    @staticmethod
+    def backward(ctx, grad_output):
+        feature_size = ctx.feature_size
+        spatial_scale = ctx.spatial_scale
+        num_orientations = ctx.num_orientations
+        clockwise = ctx.clockwise
+        num_samples = ctx.num_samples
+        rois = ctx.saved_tensors[0]
+        assert feature_size is not None
+        batch_size, num_channels, feature_height, feature_width = feature_size
```
Comment (ZwwWayne) [inline mmcv/ops/riroi_align_rotated.py:59]: feature_height -> feature_h, feature_width -> feature_w as you use out_w/out_h below.

**Pull request merged**